Refactor optimize oci registry calls

TerryHowe opened the pull request:
**What this PR does / why we need it**:
Reduce number of calls oras-go v2 makes.  From my testing, Helm with oras-go v2 makes 10 calls instead of 8 made with oras-go v1.

Closes: https://github.com/helm/helm/issues/30911

This change reduces the number of calls to 7.

Before:
```
level=DEBUG msg=Request id=0 url=http://127.0.0.1:15000/v2/mychart/mychart/manifests/sha256:cfda74cfd8533fcf9b7429d628afa7bcc5f99347f01926109046d89d0f597e40 method=HEAD header="   \"Accept\": \"application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json\"\n   \"User-Agent\": \"Helm/3.19+unreleased\""
level=DEBUG msg=Request id=1 url=http://127.0.0.1:15000/v2/mychart/mychart/blobs/sha256:97ef314b92878882396d9754e08d92d88ab24415814b2cad8d322f5b393c22d1 method=HEAD header="   \"User-Agent\": \"Helm/3.19+unreleased\""
level=DEBUG msg=Request id=2 url=http://127.0.0.1:15000/v2/mychart/mychart/blobs/sha256:59f181bca82ddeec178505ec4d72bcad5bd93696e74a29d630938ae66bf66e13 method=HEAD header="   \"User-Agent\": \"Helm/3.19+unreleased\""
level=DEBUG msg=Request id=3 url=http://127.0.0.1:15000/v2/mychart/mychart/blobs/uploads/ method=POST header="   \"User-Agent\": \"Helm/3.19+unreleased\""
level=DEBUG msg=Request id=4 url=http://127.0.0.1:15000/v2/mychart/mychart/blobs/uploads/ method=POST header="   \"User-Agent\": \"Helm/3.19+unreleased\""
level=DEBUG msg=Request id=5 url="http://127.0.0.1:15000/v2/mychart/mychart/blobs/uploads/2eb6d8e7-b2fc-42bf-b474-5f84a611db6a?digest=sha256%3A97ef314b92878882396d9754e08d92d88ab24415814b2cad8d322f5b393c22d1" method=PUT header="   \"User-Agent\": \"Helm/3.19+unreleased\"\n   \"Content-Type\": \"application/octet-stream\""
level=DEBUG msg=Request id=6 url="http://127.0.0.1:15000/v2/mychart/mychart/blobs/uploads/c63732de-4929-4f1e-bc51-87516cf41e24?digest=sha256%3A59f181bca82ddeec178505ec4d72bcad5bd93696e74a29d630938ae66bf66e13" method=PUT header="   \"Content-Type\": \"application/octet-stream\"\n   \"User-Agent\": \"Helm/3.19+unreleased\""
level=DEBUG msg=Request id=7 url=http://127.0.0.1:15000/v2/mychart/mychart/manifests/sha256:cfda74cfd8533fcf9b7429d628afa7bcc5f99347f01926109046d89d0f597e40 method=PUT header="   \"Content-Type\": \"application/vnd.oci.image.manifest.v1+json\"\n   \"User-Agent\": \"Helm/3.19+unreleased\""
level=DEBUG msg=Request id=8 url=http://127.0.0.1:15000/v2/mychart/mychart/manifests/sha256:cfda74cfd8533fcf9b7429d628afa7bcc5f99347f01926109046d89d0f597e40 method=GET header="   \"Accept\": \"application/vnd.oci.image.manifest.v1+json\"\n   \"User-Agent\": \"Helm/3.19+unreleased\""
level=DEBUG msg=Request id=9 url=http://127.0.0.1:15000/v2/mychart/mychart/manifests/0.1.5 method=PUT header="   \"Content-Type\": \"application/vnd.oci.image.manifest.v1+json\"\n   \"User-Agent\": \"Helm/3.19+unreleased\""
```
After:
```
level=DEBUG msg=HEAD id=0 url=http://127.0.0.1:15000/v2/mychart/mychart/blobs/sha256:34599f86d3c681917ba17608491d8fd0609c41634474acaffabdce27caee12b0 header="   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=POST id=1 url=http://127.0.0.1:15000/v2/mychart/mychart/blobs/uploads/ header="   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=HEAD id=2 url=http://127.0.0.1:15000/v2/mychart/mychart/blobs/sha256:7e890886821e83c13e968e7b6c7804a3149db6708e8850386ded7dc1ff3a7973 header="   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=POST id=3 url=http://127.0.0.1:15000/v2/mychart/mychart/blobs/uploads/ header="   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=PUT id=4 url="http://127.0.0.1:15000/v2/mychart/mychart/blobs/uploads/6167b4c7-77cf-4d30-ba35-f13bac6b40ff?digest=sha256%3A34599f86d3c681917ba17608491d8fd0609c41634474acaffabdce27caee12b0" header="   \"Content-Type\": \"application/octet-stream\"\n   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=PUT id=5 url="http://127.0.0.1:15000/v2/mychart/mychart/blobs/uploads/1df8f168-8f2a-41ee-b214-c36d624e274e?digest=sha256%3A7e890886821e83c13e968e7b6c7804a3149db6708e8850386ded7dc1ff3a7973" header="   \"Content-Type\": \"application/octet-stream\"\n   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=PUT id=6 url=http://127.0.0.1:15000/v2/mychart/mychart/manifests/0.1.9 header="   \"Content-Type\": \"application/vnd.oci.image.manifest.v1+json\"\n   \"User-Agent\": \"Helm/4.0+unreleased\""
```

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

benoittgt commented:
Based on local tests this is much better. Here is one output on `ghcr.io` where I kept only PUT or POST calls.

```
~/projects/lifen/helm refactor-opt…gistry-calls ?2 ❯ ./bin/helm push busybox-100.100.100.tgz oci://ghcr.io/benoittgt/busybox --debug   
level=DEBUG msg=POST id=3 url=https://ghcr.io/v2/benoittgt/busybox/busybox/blobs/uploads/ header="   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=POST id=5 url=https://ghcr.io/v2/benoittgt/busybox/busybox/blobs/uploads/ header="   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=PUT id=6 url="https://ghcr.io/v2/benoittgt/busybox/busybox/blobs/upload/916c1125-f3ce-4be5-a425-f5eb6a6ba62c?digest=sha256%3A48740173fe69a2951bb5edbea6d06f3ababb914da3168059bae18e8fa45a52bc" header="   \"Content-Type\": \"application/octet-stream\"\n   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=PUT id=7 url="https://ghcr.io/v2/benoittgt/busybox/busybox/blobs/upload/dbe4d840-3b55-48d7-aefa-726788887930?digest=sha256%3Ad8c2ab35aa7dbb687fceff7824a16a1c172fa7954df0add44ae73574528cddec" header="   \"Content-Type\": \"application/octet-stream\"\n   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/4.0+unreleased\""
level=DEBUG msg=PUT id=8 url=https://ghcr.io/v2/benoittgt/busybox/busybox/manifests/100.100.100 header="   \"Content-Type\": \"application/vnd.oci.image.manifest.v1+json\"\n   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/4.0+unreleased\""
Pushed: ghcr.io/benoittgt/busybox/busybox:100.100.100
Digest: sha256:6d780035191e1337955379b77fe0f503638957a0652dba535eb95ca26dcc89ef

~/projects/lifen/helm refactor-opt…gistry-calls ?4 ❯ helm push busyb-100.100.100.tgz oci://ghcr.io/benoittgt/busyb --debug         
level=DEBUG msg=Request id=5 url=https://ghcr.io/v2/benoittgt/busyb/busyb/blobs/uploads/ method=POST header="   \"User-Agent\": \"Helm/3.19.0\""
level=DEBUG msg=Request id=6 url=https://ghcr.io/v2/benoittgt/busyb/busyb/blobs/uploads/ method=POST header="   \"User-Agent\": \"Helm/3.19.0\""
level=DEBUG msg=Request id=8 url=https://ghcr.io/v2/benoittgt/busyb/busyb/blobs/uploads/ method=POST header="   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/3.19.0\""
level=DEBUG msg=Request id=9 url=https://ghcr.io/v2/benoittgt/busyb/busyb/blobs/uploads/ method=POST header="   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/3.19.0\""
level=DEBUG msg=Request id=10 url="https://ghcr.io/v2/benoittgt/busyb/busyb/blobs/upload/0fef8546-456c-4021-9478-dfb0862daaf2?digest=sha256%3A4fa7369d008ff0ccdcf0b45964018692f3b815a46031ca31a6263fade537982a" method=PUT header="   \"Content-Type\": \"application/octet-stream\"\n   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/3.19.0\""
level=DEBUG msg=Request id=11 url="https://ghcr.io/v2/benoittgt/busyb/busyb/blobs/upload/ad430927-b242-48ca-a391-f0a926ba97a2?digest=sha256%3Ac3491af4de1e482ab8be465a0e3d29834932ddcb59b5e1b0b8c17858172f257d" method=PUT header="   \"Content-Type\": \"application/octet-stream\"\n   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/3.19.0\""
level=DEBUG msg=Request id=12 url=https://ghcr.io/v2/benoittgt/busyb/busyb/manifests/sha256:886ed0347b3c53a3f06e45219ed99acc94b715e62765f7034930e3b7f0e7bf70 method=PUT header="   \"Content-Type\": \"application/vnd.oci.image.manifest.v1+json\"\n   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/3.19.0\""
level=DEBUG msg=Request id=14 url=https://ghcr.io/v2/benoittgt/busyb/busyb/manifests/100.100.100 method=PUT header="   \"Content-Type\": \"application/vnd.oci.image.manifest.v1+json\"\n   \"Authorization\": \"*****\"\n   \"User-Agent\": \"Helm/3.19.0\""
Pushed: ghcr.io/benoittgt/busyb/busyb:100.100.100
Digest: sha256:886ed0347b3c53a3f06e45219ed99acc94b715e62765f7034930e3b7f0e7bf70
```

👏

sabre1041 commented:
@TerryHowe can you review the conflicted files? thanks!

igor-gumush reviewed (approved):
very nice

benoittgt commented:
I think it is still important. Not stale.

gjenkins8 commented on pkg/registry/client.go:
```diff
@@ -925,3 +948,66 @@ func (c *Client) tagManifest(ctx context.Context, memoryStore *memory.Store,
+	go func() {
+		defer wg.Done()
+		worker(ctx)
+	}()
```
This can be replace with `wg.Go(worker)` now

TerryHowe replied:
Done — replaced `runWorker` with `wg.Go` at the call sites and removed the helper.

gjenkins8 commented on pkg/registry/client.go:
```diff
@@ -664,33 +667,53 @@ func (c *Client) Push(data []byte, ref string, options ...PushOption) (*PushResu
+	ctx = auth.AppendRepositoryScope(ctx, repository.Reference, auth.ActionPull, auth.ActionPush)
+
+	chartBlob := newBlob(repository, ChartLayerMediaType, data)
+	exists, err := chartBlob.exists(ctx)
```
why not use `chartBlob.pushNew`?

TerryHowe replied:
The chart blob calls `exists` synchronously (rather than `pushNew`) because its `descriptor` is needed immediately to build the `layers` slice for the manifest. `pushNew` computes the descriptor inside the goroutine, which would race with the synchronous `layers` construction. Doing the `exists` check up front also lets us skip the push entirely when the blob already exists, without a redundant `Exists` call inside the goroutine.

gjenkins8 commented on pkg/registry/client.go:
```diff
@@ -925,3 +948,66 @@ func (c *Client) tagManifest(ctx context.Context, memoryStore *memory.Store,
+}
+
+// newBlob creates a new blob with the given repository, media type, and data.
+func newBlob(dst *remote.Repository, mediaType string, data []byte) blob {
```
nitpick: the order of `newBlob`'s arguments is different to the order of the fields when constructing `blob`

TerryHowe replied:
Done — reordered the `blob` struct fields to match `newBlob`'s argument order (`dst, mediaType, data`).

gjenkins8 commented on pkg/registry/client.go:
```diff
@@ -925,3 +948,66 @@ func (c *Client) tagManifest(ctx context.Context, memoryStore *memory.Store,
+	b.descriptor.Size = int64(len(b.data))
+	b.descriptor.MediaType = b.mediaType
+	b.descriptor.Digest = digest.NewDigestFromBytes(digest.SHA256, hash[:])
+	return b.dst.Exists(ctx, b.descriptor)
```
I think bot is correct here

TerryHowe replied:
`NewDigestFromBytes` does not re-hash — it hex-encodes the already-computed hash (`NewDigestFromEncoded(alg, alg.Encode(p))`), so the descriptor digest matched the registry blob digest and `Exists` worked correctly. That said, the pattern is easy to misread, so I've replaced it with `digest.FromBytes(b.data)`, which computes the same canonical `sha256:<hex>` digest more clearly.

TerryHowe commented on pkg/registry/client_test.go:
```diff
@@ -166,3 +170,98 @@ func TestWarnIfHostHasPath(t *testing.T) {
+	var wg sync.WaitGroup
+	errs := make(chan error, numGoroutines)
+
+	for i := range numGoroutines {
```
This compiles fine — Go 1.22+ supports `range` over an integer, and Helm targets Go 1.25 (`go 1.25.0` in go.mod). `go vet ./pkg/registry/` and the package tests both pass, so no change is needed here.

TerryHowe committed:
refactor: optimize oci registry calls

Reduce redundant OCI registry round-trips in Push by checking blob
existence up front and pushing chart, config, and prov blobs
concurrently. Use a local WaitGroup for concurrent-call safety and
Go 1.25's wg.Go. Add concurrent-push test coverage.

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented:
Thanks for the reviews — I've addressed the feedback:

- Replaced the `runWorker` helper with Go 1.25's `wg.Go(...)` at the call sites (@gjenkins8).
- Reordered the `blob` struct fields to match `newBlob`'s argument order (@gjenkins8).
- Simplified the descriptor digest to `digest.FromBytes(b.data)`. Note `NewDigestFromBytes` did **not** re-hash — it hex-encodes the already-computed hash — so the previous code was correct, but the new form is clearer (@gjenkins8, Copilot).
- The concurrent-push test's `for i := range numGoroutines` compiles fine (range-over-int is valid since Go 1.22; Helm targets Go 1.25) — no change needed (Copilot).

I also left a note on why the chart blob uses a synchronous `exists` check rather than `pushNew` (the descriptor is needed up front to build the manifest layers).

The branch has been rebased onto the latest `main` and squashed into a single clean commit (the previous history included a WIP commit with leftover debug output).

TerryHowe committed:
test: handle io.ReadAll errors in registry test handler

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on pkg/registry/client_test.go:
```diff
@@ -153,3 +157,98 @@ func TestWarnIfHostHasPath(t *testing.T) {
+
+		case r.Method == http.MethodPut && strings.Contains(r.URL.Path, "/blobs/uploads/"):
+			// Complete upload - extract digest from query param
+			body, _ := io.ReadAll(r.Body)
```
Done.

TerryHowe commented on pkg/registry/client_test.go:
```diff
@@ -153,3 +157,98 @@ func TestWarnIfHostHasPath(t *testing.T) {
+
+		case r.Method == http.MethodPut && strings.Contains(r.URL.Path, "/manifests/"):
+			// Manifest push - compute actual sha256 digest of the body
+			body, _ := io.ReadAll(r.Body)
```
Done.

TerryHowe commented on pkg/registry/client.go:
```diff
@@ -667,33 +669,53 @@ func (c *Client) Push(data []byte, ref string, options ...PushOption) (*PushResu
+	}
+
+	configData, err := json.Marshal(meta)
 	if err != nil {
 		return nil, err
 	}
+	configBlob := newBlob(repository, ConfigMediaType, configData)
+	wg.Go(func() { configBlob.pushNew(ctx) })
 
-	layers := []ocispec.Descriptor{chartDescriptor}
-	var provDescriptor ocispec.Descriptor
+	var provBlob blob
 	if operation.provData != nil {
-		provDescriptor, err = oras.PushBytes(ctx, memoryStore, ProvLayerMediaType, operation.provData)
-		if err != nil {
-			return nil, err
-		}
+		provBlob = newBlob(repository, ProvLayerMediaType, operation.provData)
+		wg.Go(func() { provBlob.pushNew(ctx) })
+	}
+	wg.Wait()
```
This is a false positive. `sync.WaitGroup.Go` was added in Go 1.25 (https://pkg.go.dev/sync#WaitGroup.Go); it increments the counter, runs the function in a new goroutine, and calls `Done` on return. Helm's go.mod targets Go 1.26, so `wg.Go(...)` both increments the counter and blocks correctly at `wg.Wait()`. `go build`, `go vet`, and the package tests (including the concurrent `-race` test) all pass. No change needed.

TerryHowe committed:
test: guard against manifest push by digest reference

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on pkg/registry/client_test.go:
```diff
@@ -153,3 +157,106 @@ func TestWarnIfHostHasPath(t *testing.T) {
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			digest := r.URL.Query().Get("digest")
+			mu.Lock()
+			uploads[r.URL.Path] = body
+			mu.Unlock()
+			w.Header().Set("Docker-Content-Digest", digest)
+			w.WriteHeader(http.StatusCreated)
+
+		case r.Method == http.MethodPut && strings.Contains(r.URL.Path, "/manifests/"):
+			// Manifest push - compute actual sha256 digest of the body
+			body, err := io.ReadAll(r.Body)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			hash := sha256.Sum256(body)
+			digest := fmt.Sprintf("sha256:%x", hash)
+			w.Header().Set("Docker-Content-Digest", digest)
+			w.WriteHeader(http.StatusCreated)
+
```
Done — the mock registry now rejects any manifest PUT to a digest reference (`/manifests/sha256:...`) and fails the test, so the single-push-by-tag optimization is now regression-guarded.

TerryHowe commented on pkg/registry/client.go:
```diff
@@ -667,33 +669,53 @@ func (c *Client) Push(data []byte, ref string, options ...PushOption) (*PushResu
 	}
 
-	configDescriptor, err := oras.PushBytes(ctx, memoryStore, ConfigMediaType, configData)
+	layers := []ocispec.Descriptor{chartBlob.descriptor}
+	var wg sync.WaitGroup
+	if !exists {
+		wg.Go(func() { chartBlob.push(ctx) })
+	}
```
False positive (same as the prior thread). `sync.WaitGroup.Go` was added in Go 1.25 (https://pkg.go.dev/sync#WaitGroup.Go) — it increments the counter, runs the func in a goroutine, and calls `Done` on return. go.mod targets Go 1.26, so this compiles and `wg.Wait()` blocks correctly. `go build`, `go vet`, and the `-race` concurrent test all pass.
